infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -383,28 +438,39 @@ public boolean isServerReadyToHandleMsg(CorfuMsg msg) {
     private synchronized void resetLogUnit(CorfuPayloadMsg<Long> msg,
                                            ChannelHandlerContext ctx, IServerRouter r) {
 
+        if (msg.getPayload() <= serverContext.getLogUnitEpochWaterMark()
+                || msg.getPayload() != serverContext.getServerEpoch()) {
+            log.info("LogUnit Server Reset request received but reset already done.");
+            r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+            return;
+        }
+
         // Check if the reset request is with an epoch greater than the last reset epoch seen to
-        // prevent multiple reset in the same epoch. and should be equal to the current router
+        // prevent multiple reset in the same epoch and should be equal to the current router
         // epoch to prevent stale reset requests from wiping out the data.
-        if (msg.getPayload() > serverContext.getLogUnitEpochWaterMark()
-                && msg.getPayload() == serverContext.getServerEpoch()) {
-            serverContext.setLogUnitEpochWaterMark(msg.getPayload());
-            batchWriter.addTask(RESET, msg)
+        serverContext.setLogUnitEpochWaterMark(msg.getPayload());
+
+        Lock lock = resetLock.writeLock();
+        lock.lock();
+
+        try {
+            CompletableFuture<Void> cf = batchWriter
+                    .addTask(RESET, msg)
                     .thenRun(() -> {
                         dataCache.invalidateAll();
                         log.info("LogUnit Server Reset.");
                         r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
-                    }).exceptionally(ex -> {
-                handleException(ex, ctx, msg, r);
-                return null;
-            });
-        } else {
-            log.info("LogUnit Server Reset request received but reset already done.");
-            r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+                    })
+                    .exceptionally(ex -> {
+                        handleException(ex, ctx, msg, r);
+                        return null;
+                    });
+            cf.join();
```

Comment:
Can you put a comment to indicate that this has to be a synchronous call for the lock to take effect.

Reply:
done